velox/vector/VariantToVector.cpp

```diff
@@ -77,4 +77,165 @@ ArrayVectorPtr variantArrayToVector(
+    velox::memory::MemoryPool* pool) {
+  switch (counter.kind) {
+    case TypeKind::ARRAY: {
+      BufferPtr sizes =
+          AlignedBuffer::allocate<vector_size_t>(counter.elementCount, pool, 0);
+      BufferPtr offsets =
+          AlignedBuffer::allocate<vector_size_t>(counter.elementCount, pool, 0);
+      BufferPtr nulls =
+          AlignedBuffer::allocate<vector_size_t>(counter.elementCount, pool, 0);
+      VectorPtr elements = allocateVector(counter.children.at(0), pool);
+      return std::make_shared<ArrayVector>(
+          pool,
+          ARRAY(elements->type()),
+          nulls,
+          counter.elementCount,
+          std::move(offsets),
+          std::move(sizes),
+          std::move(elements));
+    }
+    default: {
+      TypePtr type;
+      type = createScalarType(counter.kind);
+      return BaseVector::create(std::move(type), counter.elementCount, pool);
+    }
+  }
+}
+
+template <TypeKind Kind>
+void setElementInFlatVector(
+    vector_size_t idx,
+    const variant& v,
+    VectorPtr& vector) {
+  using NativeType = typename TypeTraits<Kind>::NativeType;
+  auto asFlat = vector->asFlatVector<NativeType>();
+  asFlat->set(idx, NativeType{v.value<NativeType>()});
+}
+
+static void incrementChildElementsInserted(NestedTypeCounter& counter) {
+  for (NestedTypeCounter& child : counter.children) {
+    incrementChildElementsInserted(child);
+    child.elementsInserted += 1;
+  }
+}
+
+static void insertVariantIntoVector(
+    NestedTypeCounter& counter,
+    const variant& v,
+    VectorPtr& vector) {
+  if (v.isNull()) {
+    vector->setNull(counter.elementsInserted, true);
+    if (counter.kind == TypeKind::ROW) {
+      incrementChildElementsInserted(counter);
+    }
+  } else
+    switch (counter.kind) {
+      case TypeKind::ARRAY: {
+        auto asArray = vector->as<ArrayVector>();
+        const std::vector<variant>& elements = v.array();
+        vector_size_t offset = 0;
+        if (counter.elementsInserted != 0) {
+          offset = asArray->offsetAt(counter.elementsInserted - 1) +
+              asArray->sizeAt(counter.elementsInserted - 1);
+        }
+        asArray->setOffsetAndSize(
+            counter.elementsInserted, offset, elements.size());
+        for (const variant& elt : elements) {
+          insertVariantIntoVector(
+              counter.children.at(0), elt, asArray->elements());
+        }
+        break;
+      }
+      default: {
+        VELOX_DYNAMIC_SCALAR_TYPE_DISPATCH(
+            setElementInFlatVector,
+            counter.kind,
+            counter.elementsInserted,
+            v,
+            vector);
+        break;
+      }
+    }
+  counter.elementsInserted += 1;
+}
+
+VectorPtr variantsToVector(
+    const std::vector<variant>& variants,
+    velox::memory::MemoryPool* pool) {
+  NestedTypeCounter counter;
+  for (const variant& v : variants) {
+    computeElementsCountsAndType(counter, v);
+  }
+  VectorPtr resultVector = allocateVector(counter, pool);
```

Comment:
Instead of using allocateVector, its better to just use BaseVector::create(type..) which takes a type and constructs the whole vector tree.

Reply:
Using `BaseVector::create` now, thanks for the suggestion!